examples/scene/prefab.rs

```diff
@@ -0,0 +1,123 @@
+
+#[derive(Reflect, Clone)]
+#[reflect(Component)]
+struct A {
+    message: String,
+}
+
+impl Default for A {
+    fn default() -> Self {
+        A { message: "Hello!".to_string() }
+    }
+}
+#[derive(Reflect, Clone, Default)]
+#[reflect(Component)]
+struct B {
+    data: Vec<usize>,
+}
+
+fn locate_prefab(
+    query: Query<(&A, &B), (Added<A>, With<TrackingComponent>)>,
+) {
+    for (comp_a, comp_b) in query.iter() {
+        println!("{}, {:?}", comp_a.message, comp_b.data);
+    }
+}
+
+fn main() {
+    let mut app = App::build();
+    app
+        .init_resource::<PrefabFactory>()
+        .add_system_to_stage(
+            CoreStage::Last, 
+            prefab_factory_system_ex.exclusive_system()
+                .with_run_criteria(run_if_queue_occupied.system())
+        )
+        .add_startup_system(setup.system())
+    ;
+
+    app.run()
+}
+
+
+
+#[derive(Default)]
+pub struct PrefabFactory {
+    queue: HashMap<Entity, Handle<DynamicScene>>,
+}
+
+impl PrefabFactory {
+    pub fn add_to_queue(&mut self, entity: Entity, handle: Handle<DynamicScene>) {
+        self.queue.insert(entity, handle);
+    }
+}
+// Run condition to prevent the factory from running while empty.
+pub fn run_if_queue_occupied(
+    prefab_factory: Res<PrefabFactory>
+) -> ShouldRun {
+    match !prefab_factory.queue.is_empty() {
+        false => ShouldRun::No,
+        true => ShouldRun::Yes,
+    }
+}
+
+pub fn prefab_factory_system_ex(world: &mut World) {
+    world.resource_scope(|world, mut factory: Mut<PrefabFactory>| {
+        let registry = world.get_resource::<TypeRegistry>().unwrap().clone();
+        let type_registry = registry.read();
+        let mut complete = Vec::default();
+
+        world.resource_scope(|world, dynamic_scenes: Mut<Assets<DynamicScene>>| {
+            for (entity, handle) in factory.queue.iter() {
+                if let Some(scene) = dynamic_scenes.get(handle) {
+                    for component in scene.entities.first().unwrap().components.iter() {
+                        let registration = type_registry
+                            .get_with_name(component.type_name())
+                            .ok_or_else(|| SceneSpawnError::UnregisteredType {
+                                type_name: component.type_name().to_string(),
+                            }).unwrap();
+                        let reflect_component =
+                            registration.data::<ReflectComponent>().ok_or_else(|| {
+                                SceneSpawnError::UnregisteredComponent {
+                                    type_name: component.type_name().to_string(),
+                                }
+                            }).unwrap();
+                        if world
+                            .entity(*entity)
+                            .contains_type_id(registration.type_id())
+                        {
+                            reflect_component.apply_component(world, *entity, &**component);
```

Comment:
We should be able to avoid some of this ugly dereferencing by destructuring when these variables are declared at the start of each for loop.

Reply:
E.g. `for (&entity, handle) in...`